packages/perseus-core/src/widgets/upgrade.ts

```diff
@@ -0,0 +1,158 @@
+    // figure that out before depending solely on types.
+    if (!_.isString(type)) {
+        throw new PerseusError(
+            "widget type must be a string, but was: " + type,
+            Errors.Internal,
+        );
+    }
+
+    if (!isWidgetRegistered(type)) {
+        // If we have a widget that isn't registered, we can't upgrade it
+        // TODO(aria): Figure out what the best thing to do here would be
+        return oldWidgetInfo;
+    }
+
+    // Unversioned widgets (pre-July 2014) are all implicitly 0.0
+    const initialVersion = oldWidgetInfo.version || {major: 0, minor: 0};
+    const latestVersion = getCurrentVersion(type);
+
+    // If the widget version is later than what we understand (major
+    // version is higher than latest, or major versions are equal and minor
+    // version is higher than latest), don't perform any upgrades.
+    if (
+        initialVersion.major > latestVersion.major ||
+        (initialVersion.major === latestVersion.major &&
+            initialVersion.minor > latestVersion.minor)
+    ) {
+        return oldWidgetInfo;
+    }
+
+    // We do a clone here so that it's safe to mutate the input parameter
+    // in propUpgrades functions (which I will probably accidentally do at
+    // some point, and we would like to not break when that happens).
+    let newEditorOptions = _.clone(oldWidgetInfo.options) || {};
+
+    const upgradePropsMap = getWidgetOptionsUpgrades(type);
+
+    // Empty props usually mean a newly created widget by the editor,
+    // and are always considerered up-to-date.
+    // Mostly, we'd rather not run upgrade functions on props that are
+    // not complete.
+    if (_.keys(newEditorOptions).length !== 0) {
+        // We loop through all the versions after the current version of
+        // the loaded widget, up to and including the latest version of the
+        // loaded widget, and run the upgrade function to bring our loaded
+        // widget's props up to that version.
+        // There is a little subtlety here in that we call
+        // upgradePropsMap[1] to upgrade *to* version 1,
+        // (not from version 1).
+        for (
+            let nextVersion = initialVersion.major + 1;
+            nextVersion <= latestVersion.major;
+            nextVersion++
+        ) {
+            if (upgradePropsMap[String(nextVersion)]) {
+                newEditorOptions =
+                    upgradePropsMap[String(nextVersion)](newEditorOptions);
+            } else {
+                throw new PerseusError(
+                    "No upgrade found for widget. Cannot render.",
+                    Errors.Internal,
+                    {
+                        metadata: {
+                            type,
+                            fromMajorVersion: nextVersion - 1,
+                            toMajorVersion: nextVersion,
+                        },
+                    },
+                );
```

Comment:
Here's an important change. Originally we just `Log`ged and continued. We don't have `Log` in PerseusCore so I wasn't sure what to do. IMO:

- I feel like it's unlikely we'll hit a major bump that hasn't been handled by the upgrader
- If we're worried about it, we could write tests or linters to make sure we don't do a major bump without an upgrader
- Upgrading might disappear with Ben's current parsing work

So that's why I did what I did. Very open to ideas.

Reply:
Done